File: resteasy-reactive/src/main/java/org/example/PostResources.java
package org.example;

import io.smallrye.mutiny.Uni;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.validation.Valid;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;

import java.net.URI;
import java.util.UUID;
import java.util.logging.Logger;

import static jakarta.ws.rs.core.Response.*;

@Path("/posts")
@ApplicationScoped
public class PostResources {

    private final static Logger LOGGER = Logger.getLogger(PostResources.class.getName());

    private final PostRepository posts;

    @Inject
    public PostResources(PostRepository posts) {
        this.posts = posts;
    }

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public Uni<Response> getAllPosts() {
        return this.posts.listAll().map(data -> Response.ok(data).build());
    }

    @POST
    @Consumes(MediaType.APPLICATION_JSON)
    public Uni<Response> savePost(@Valid CreatePostRequest formData) {
        var post = Post.builder()
                .title(formData.title())
                .content(formData.content())
                .build();
        return this.posts.persist(post)
                .map(saved -> created(URI.create("/posts/" + saved.id)).build());
    }

    @Path("{id}")
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public Uni<Response> getPostById(@PathParam("id") final UUID id) {
        return this.posts.findById(id)
                .map(data -> {
                    if (data == null) {
                        return null;
                    }
                    return ok(data).build();
                })
                .onItem().ifNull().continueWith(status(Status.NOT_FOUND).build());
        //.onFailure(PostNotFoundException.class).recoverWithItem(status(Status.NOT_FOUND).build());
    }

    @Path("{id}")
    @PUT
    @Consumes(MediaType.APPLICATION_JSON)
    public Uni<Response> updatePost(@PathParam("id") final UUID id, @Valid UpdatePostRequest updateForm) {
        return this.posts.findById(id)
                .onItem().ifNull().failWith(() -> new PostNotFoundException(id))
                .map(data -> Post.builder().id(id).title(updateForm.title()).content(updateForm.content()).build())
                .flatMap(this.posts::update)
                .map(updated -> updated > 0 ? Status.NO_CONTENT : Status.NOT_FOUND)
                .map(status -> status(status).build())
                .onFailure(PostNotFoundException.class).recoverWithItem(status(Status.NOT_FOUND).build());
    }

    @DELETE
    @Path("{id}")
    public Uni<Response> delete(@PathParam("id") UUID id) {
        return this.posts.deleteById(id)
                .map(deleted -> deleted ? Status.NO_CONTENT : Status.NOT_FOUND)
                .map(status -> status(status).build());
    }
}

File: rest-kotlin-co/src/main/kotlin/com/example/demo/PostResources.kt
package com.example.demo

import io.smallrye.mutiny.coroutines.awaitSuspending
import jakarta.enterprise.context.ApplicationScoped
import jakarta.validation.Valid
import jakarta.ws.rs.*
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import java.net.URI
import java.util.*
import java.util.logging.Logger

@ApplicationScoped
@Path("/posts")
class PostResources(private val posts: PostRepository) {
    companion object {
        private val LOGGER = Logger.getLogger(PostResources::class.java.name)
    }

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    suspend fun getAllPosts(): Response {
        val data = posts.listAll().awaitSuspending()
        return Response.ok(data).build()
    }

    @POST
    @Consumes(MediaType.APPLICATION_JSON)
    suspend fun savePost(formData: @Valid CreatePostCommand): Response {
        val post = Post(
            title = formData.title,
            content = formData.content
        )
        posts.persist(post).awaitSuspending()
        return Response.created(URI.create("/posts/" + post.id)).build()
    }

    @Path("{id}")
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    suspend fun getPostById(@PathParam("id") id: UUID): Response {
        val post = posts.findById(id).awaitSuspending() ?: run {
            return Response.status(Response.Status.NOT_FOUND).build()
        }

        return Response.ok(post).build()

    }

    @Path("{id}")
    @PUT
    @Consumes(MediaType.APPLICATION_JSON)
    suspend fun updatePost(@PathParam("id") id: UUID, updateForm: @Valid UpdatePostCommand): Response {
        val post = posts.findById(id).awaitSuspending() ?: run {
            return Response.status(Response.Status.NOT_FOUND).build()
        }
        posts.update(post.apply {
            title = updateForm.title
            content = updateForm.content
        })

        return Response.noContent().build()
    }

    @DELETE
    @Path("{id}")
    suspend fun delete(@PathParam("id") id: UUID): Response {
        val deleted = posts!!.deleteById(id).awaitSuspending()
        return if (deleted) Response.noContent().build() else Response.status(Response.Status.NOT_FOUND).build()
    }
}
